tempesta_fw/http.c

```diff
@@ -3873,16 +3873,20 @@ static void
 tfw_http_resp_terminate(TfwHttpMsg *hm)
 {
 	TfwFsmData data;
-	int r;
+	int r = 0;
 
 	/*
-	 * Add absent message framing information. If there is no body, add
-	 *  Content-length header: it requires less modifications.
+	 * Add absent message framing information. It's possible to add a
+	 * 'Content-Length: 0' header, if the Transfer-Encoding header is not
+	 * set, but keep more generic solution and transform to chunked.
+	 * It's the only possible modification for future proxy mode.
+	 * If the framing information can't be added, then close client
+	 * connection after response is forwarded.
 	 */
-	r = (hm->body.len)
-		? tfw_http_msg_to_chunked(hm)
-		: TFW_HTTP_MSG_HDR_XFRM(hm, "Content-Length", "0",
-					TFW_HTTP_HDR_CONTENT_LENGTH, 0);
+	if (!test_bit(TFW_HTTP_B_CHUNKED_APPLIED, hm->flags))
```

Comment:
`TFW_HTTP_B_CHUNKED_APPLIED` is set only if there were `chunked` somewhere in `Transfer-Encoding`, but it wasn't the last one. So it won't be set for `Transfer-Encoding: deflate`, for example. Shouldn't we switch `TFW_HTTP_B_CONN_CLOSE` on in this case too?

Reply:
No, we needn't. We are trying our best to keep connections open if possible, it actually save our resources:
- Less handshakes TCP or more costly TLS ones,
- There can be more pipelined requests from that client, that can be sent to other connections, and we may store that request and corresponding responses. But if we chose to close the client connections all that requests and responses will be discarded. It's very likely that the client will reopen its connection and request the same resources one more time. So keeping client connection open may help us to reduce backed traffic.

In that particular case with `Transfer-Encoding: deflate` we can add our own encoding to keep the message framed with out connection close - we can encode the message to chunked encoding: `Transfer-Encoding: deflate, chunked`. The function `tfw_http_msg_to_chunked()` makes that, which called from tfw_http_resp_terminate.

If the flag `TFW_HTTP_B_CHUNKED_APPLIED` is set, then we can't encode the message to chunked encoding once again, and the only thing we can do to keep the framing - is to close the client connection.

Reply:
I also have a test case for  `Transfer-Encoding: deflate` -> `Transfer-Encoding: deflate, chunked` conversion: https://github.com/tempesta-tech/tempesta-test/pull/145/files#diff-d1f29d9179db6bfb793086c219087719R198

Reply:
Ah, now I see, I was wrong. Thanks for the clarifications.